### DISCUSSION TITLE
Assign an Image to a Windows Bitmap

### DISCUSSION THREAD

**Discussion by michael-scott-phoseon** [libvips/libvips#2465] [Q&A]
I can now load an image using Image.Tiffload(filename, pageno).  

But now I need to assign that image to either a System.Drawing.Bitmap (Windows Forms), or better yet, a System.Windows.Media.ImageSource or System.Windows.Media.BitmapSource  (WPF - preferred), so I can display it in a window.  I can't find any samples for that in either libvips or netvips.

**Comment by michael-scott-phoseon** [comment]
I am trying to insert the libvip Image into a memory stream but am still having issues.

Our images are in a pyramidal ome-tiff file where the images are are of 81920 x 98304, so using JpegsaveStream won't work, as it has a 65500 x 65500 size limit. So we are trying to use Jp2ksaveSteam instead.  However, it throws this error:

          NetVips.VipsException: 'no such operation jp2ksave_target VipsOperation: class "jp2ksave_target" not found

Does libvips and / or NetVips not support Jpeg 2000?  We are trying to insert a libvips Image into a WPF Window, so are trying to move the stream into a BitmapSource.  But I am even willing to place it into a Windows Forms Bitmap. Here is the method I'm using:

        private static BitmapImage NetVipsToImageSource(NetVips.Image img)
        {
            using (MemoryStream memory = new MemoryStream())
            {
                img.Jp2ksaveStream(memory);
                //img.JpegsaveStream(memory);
                //img.WriteToStream(memory, ".jpg");
                memory.Position = 0;
                BitmapImage bitmapImage = new BitmapImage();
                bitmapImage.BeginInit();
                bitmapImage.StreamSource = memory;
                bitmapImage.CacheOption = BitmapCacheOption.OnLoad;
                bitmapImage.EndInit();

                return bitmapImage;
            }
        }

Am I doing something wrong?  Is Jpeg 2000 not supported?  Is there another way to do this?

Michael

**Reply by jcupitt** [reply]
Hi, jpeg2000 support was added in 8.11. Perhaps you are using an older libvips?

However, taking the entire image fom netvips to a drawable via jp2k will be extremely slow and need a huge amount of memory.

You can do it directly -- just wrap a bitmap around a memory area from netvips. You'll need to lock the memory area to prevent the GC from moving it.

Even better: only convert the area you need for screen display, and trigger further conversion as the user pans and zooms.

**Comment by kleisauke** [comment]
Answered here: https://github.com/kleisauke/net-vips/issues/147#issuecomment-939282359.

**Reply by jcupitt** [reply]
Please don't cross-post questions @michael-scott-phoseon, it doubles the work for us.

**Comment by michael-scott-phoseon** [comment]
John,

First of all, I can not express my gratitude for the help you have provided already.  Prior to 3wks ago, I’d never even HEARD of OME-TIFF, and your help has been invaluable. I’m almost there, so I should be out of your hair soon.

According to NuGet, I am using “NetVips.Native.win-x64”, which is version 8.11.4, and the Description claims that it is “Native binaries of libvips for Windows (x64)”.  And our OS must be 64bit Windows.

I also have the following NuGet Packages:

     NetVips 2.0.1 : .NET bindings for libvips (though I suspect this may be a 32bit version of NetVips.Native.winx64 so may not need it, but am unsure)
     NetVips.Extensions 2.0.0: a set if extension methods and helpers for NetVips (heard a rumor that it contains a .ToBitmap method which may solve one of my other problems)
     NetVips.Native.win-64 8.11.4 (I suspect this replaces the need for NetVips 2.0.1)

I’m going to check out this .ToBitmap idea, and if it doesn’t work, I will definitely give your suggestions a try

**Comment by michael-scott-phoseon** [comment]
John,

My apologies. It won’t happen again.  And seriously, thank you so much for all the help you’ve provided the past 2 weeks.

Michael
